activitysim/abm/tables/shadow_pricing.py

```diff
@@ -519,15 +733,108 @@ def update_shadow_prices(self):
 
             new_shadow_prices = self.shadow_prices + adjustment
 
+        elif shadow_price_method == "simulation":
+            # - NewMethod
+            """
+            C_j = (emp_j/sum(emp_j))/(workers_j/sum(workers_j))
+
+            if C_j > 1: #under-estimate workers in zone
+
+                shadow_price_j = 0
+
+            elif C_j < 1: #over-estimate workers in zone
+
+                shadow_price_j = -999
+                resimulate n workers from zone j, with n = int(workers_j-emp_j/sum(emp_j*workers_j))
+            """
+            percent_tolerance = self.shadow_settings["PERCENT_TOLERANCE"]
+            sampled_persons = pd.DataFrame()
+            persons_merged = inject.get_table("persons_merged").to_frame()
+
+            # need to join the segment to the choices to sample correct persons
+            segment_to_name_dict = self.shadow_settings.get(
+                "", default_segment_to_name_dict
+            )
+            segment_name = segment_to_name_dict[self.model_selector]
+
+            if type(self.choices_synced) != pd.DataFrame:
+                self.choices_synced = self.choices_synced.to_frame()
+
+            choices_synced = self.choices_synced.merge(
+                persons_merged[segment_name],
+                how="left",
+                left_index=True,
+                right_index=True,
+            ).rename(columns={segment_name: "segment"})
+
+            for segment in self.segment_ids:
+                desired_size = self.target[segment]
+                modeled_size = self.modeled_size[segment]
+
+                # loop over other segments and add to modeled share if they have the same target
+                for other_segment in self.segment_ids:
+                    if (segment != other_segment) & (
+                        self.target[segment].equals(self.target[other_segment])
+                    ):
+                        modeled_size = modeled_size + self.modeled_size[other_segment]
+
+                # want to match distribution, not absolute numbers so share is computed
+                desired_share = desired_size / desired_size.sum()
+                modeled_share = modeled_size / modeled_size.sum()
+
+                sprice = desired_share / modeled_share
+                sprice.fillna(0, inplace=True)
+                sprice.replace([np.inf, -np.inf], 0, inplace=True)
+
+                # shadow prices are set to -999 if overassigned or 0 if the zone still has room for this segment
+                self.shadow_prices[segment] = np.where(
```

Comment:
When I run the 25-zone shadow pricing example with the "full" test population, which is 5000 households, the model is crashing on school location iteration 2.  It seems that there are only 2 high school zones possible, and after the first iteration one is over subscribed and one is under subscribed (obviously), but it's within the `percent_tolerance` and both zones get closed at this line of code. The problem is that one high schooler still isn't assigned a school location, and with all zones closed the model then subsequently crashes.  This is a corner case but we don't want corner cases to crash the model, and while unlikely in realistic size applications it's still conceivable.

A proposed fix: add a check here to make sure that we never close *all* the zones.  You can check if the max `sprice` is less than `1 + percent_tolerance / 100` and if so, make the shadow prices such that at least that max price zone stays open.  You'll probably also want to flag this condition and then prevent additional iterations / reassignments in this segment, as you're going to be as converged as you can possibly get within the tolerances of the integer simulation.

Reply:
Good catch.  I didn't see it because I was running the "full" test population with multi-processing and it didn't fail.  Obviously this is bad because it means single process and multi-process results were different!  The problem stemmed from the sampling of people from over-subscribed zones being done with the built in pandas sample.  The sampling logic was re-written to use ActivitySim's RNG.  I have turned on shadow pricing for the small test example (looks like shadow pricing was never included in any CI tests before this!) and included a multi-process test as well.

As for the edge case you found.  Totally agree that we don't want this to crash the model.  We had some logic in location_choice.py that was trying to deal with this instance where there were no available zones, but it didn't catch the edge case where one segment was converged but other segments were not  (e.g. high school converged but not college).  The implemented solution was to treat the segment as converged if all of the zones are within tolerance and then then not sample any people for re-simulation. By dealing with this fix in the sampling step, we eliminated the need for the logic in the location_choice.py to try to catch unavailable zones.

All of these changes have been pushed and the (updated) tests are passing.

Reply:
Thanks @dhensle!  I confirm my edge case looks fixed and things are running more smoothly now.